modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -4,129 +4,192 @@ import cats.data.{EitherT, NonEmptyList, NonEmptySet}
+        EitherT
+          .fromEither(KryoSerializer[F].deserialize[DAGTransaction](hex.toBytes))
+          .leftMap(_ => MalformedTransaction)
+          .flatMap { transaction =>
+            EitherT {
+              signature.publicKey.hexBytes.toPublicKeyByEC
+                .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
+            }.flatMap { proof =>
+              EitherT
+                .fromEither(
+                  KryoSerializer[F]
+                    .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
+                )
+                .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
+                .map(Hex.fromBytes(_))
+            }
+          }
+
+      override def getPayloads(
+        operations: NonEmptyList[Operation],
+        metadataResult: MetadataResult
+      ): EitherT[F, ConstructionError, PayloadsResult] =
+        EitherT(
+          saltGen().map(transactionSalt =>
+            for {
+              (positiveOperation, negativeOperation) <- Either.fromOption(
+                getPayloadOperations(operations),
+                InvalidPayloadOperations
+              )
+
+              _ <- Either.cond(
+                positiveOperation.amount.negate.toOption === negativeOperation.amount.some,
+                (),
+                InvalidPayloadOperations
+              )
+
+              transactionAmount <- positiveOperation.amount.value.toTransactionAmount
+                .leftMap(_ => InvalidPayloadOperations)
+
+              sourceAddress = negativeOperation.account.address
+              transactionFee <-
+                metadataResult.suggestedFee match {
+                  case None      => TransactionFee.zero.asRight
+                  case Some(amt) => amt.value.toTransactionFee.leftMap(_ => InvalidSuggestedFee)
+                }
+
+              dagTransaction =
+                DAGTransaction(
+                  source = sourceAddress,
+                  destination = positiveOperation.account.address,
+                  amount = transactionAmount,
+                  fee = transactionFee,
+                  parent = metadataResult.metadata.lastReference,
+                  salt = transactionSalt
+                )
+
+              serializedTxn <-
+                KryoSerializer[F]
+                  .serialize(dagTransaction)
+                  .leftMap(e => SerializationError(e.getMessage))
+
+              txHash <-
+                serializedTxn.hash
+                  .leftMap[ConstructionError](_ => MalformedTransaction)
+                  .map(h => Hex(h.value))
+
+              payload = SigningPayload(AccountIdentifier(sourceAddress, none), txHash, SignatureType.ECDSA)
+
+            } yield PayloadsResult(Hex.fromBytes(serializedTxn), NonEmptyList.one(payload))
+          )
+        )
+
+      private def getPayloadOperations(operations: NonEmptyList[Operation]): Option[(Operation, Operation)] =
```

Comment:
I guess that we are requiring these two operations to be adding and subtracting the balance. Whereas these being a pair of a negative and positive values is being checked it is not checked whether one is the negation of the other. Not sure if that's what should be checked here just giving my 2 cents. Another thing is that if we are generalizing operations, and then assuring that in fact they are only those two where one is the adding and second subtracting maybe we can restrict the operations, to be exactly a tuple of those two? Not sure if we can do that, thinking out loud here.

Reply:
@TheMMaciek @marcinwadon 
Looking for clarification:
Do you want me to refactor so that it takes a tuple or are we good here?

Reply:
That depends on whether it interferes with rosetta. From my limited point of view I f we know that there are exactly two operations required then we can possibly express it as a tuple or a case class with two operations. That way we can have some validation work automatically based on the schema we use (no checking that the list has exactly two elements needed then). But I don't know whether it somehow interferes with Rosetta so the decision is yours.